test/controllers/api/v2/hosts_controller_test.rb

```diff
@@ -1213,7 +1213,7 @@ def setup
   test "should not update with invalid mac" do
     mac = RFauxFactory.gen_alpha
     put :update, params: { :id => @host.id, :host => {:mac => mac} }
-    assert_response :unprocessable_entity, "Can update host with invalid mac #{mac}"
+    assert_response :internal_server_error, "Can update host with invalid mac #{mac}"
```

Comment:
This looks like a bad change. Invalid data should not result in an internal server error. That's why it was using a validation.

Reply:
Sorry for the long time, I had not seen this comment. 

I agree this change don't stick with a "normal" API behavior. As Foreman uses host values _before_ validation we need to find a way to validate without throwing an exception, but silently ignoring values is a no-go too. The simplest solution is either to downcase the mac when the attribute is _set_ or when it's _get_, and leave the possibility to set incorrect values until it's validated (e.g. : in a11dd500d1f164744db1f06ab38e8b86eabe97a7)

Better solutions are welcome!

Reply:
@mtancoigne How to know what operations are made by Foreman _before_ validation?

Reply:
It's hard to tell precisely without digging deep; at first sight, [the compute profile is applied](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L141). The issue we try to address originally is the use of mixed upcased/downcased mac adresses somewhere in the steps of a virtual machine creation.

The mac addresses [being downcased](https://github.com/theforeman/foreman/blob/develop/lib/net/validations.rb#L122) [_before_ validation](https://github.com/theforeman/foreman/blob/develop/app/models/nic/base.rb#L14) lets me suppose that host values are used between instanciation and validation.

Reply:
I think you can split validation of mac and converting it to downcase into two validation methods. One that validates should raise error while the downcase should only downcase (if its a string).

Reply:
If there are validators, they won't be triggered before the validation step and the current behavior won't change. If the downcase should only downcase, the best place for it is still in the `mac=` method, so it's downcased on assignation, right ?

Reply:
Now outdated, closing conversation.